### PR TITLE
Make the file generation more transparent

### PR DESCRIPTION
**Pull request by thufschmitt**
Regenerate the files (defined in `files.*`) automatically when entering the shell (if needed).

Also unify that mechanism with the one used to generate the lockfile so that everything is generated at once.

The lockfile generation is slightly hacky because there's no easy way to pass its content from Nix to Nickel. But that'll have to do the job.

### CODE REVIEW

File: lib/lib.nix [hunk 1]
```diff
@@ -210,9 +206,14 @@
     nickelResult = callNickel {
       inherit nickelFile baseDir flakeInputs lockFileContents;
     };
+    enrichedFlakeInputs =
+      flakeInputs
+      // {
+        "%%organist_internal".nickelLock = builtins.toFile "nickel.lock.ncl" (buildLockFileContents lockFileContents);
```
Comment (yannham) [inline lib/lib.nix:212]: Could be nice to add a little comment explaining why we need to do that, for future you, me and others

File: run-test.sh [hunk 2]
```diff
@@ -120,7 +119,7 @@ test_example () (
   cp -r "$examplePath" ./example
   pushd ./example
   prepare_shell
-  nix run .\#regenerate-files --print-build-logs
+  # nix run .\#regenerate-files --print-build-logs
```
Comment (yannham) [inline run-test.sh:122]: Is this line still there on purpose or should it just be scraped entirely?

**Review by yannham — APPROVED** [review]

File: run-test.sh (hunk 2 above)
Comment (thufschmitt) [reply run-test.sh]: Nope, that's purely accidental. Removed it

File: lib/lib.nix (hunk 1 above)
Comment (thufschmitt) [reply lib/lib.nix:215]: Indeed. Fixed

**Comment by thufschmitt** [comment]
It turns out that this blows-up the Nickel evaluation time (~x30 according to my benchmarks). I'll have to investigate why before we can merge this

**Comment by yannham** [comment]
x30 for such a change is really suspicious...

**Comment by thufschmitt** [comment]
> x30 for such a change is really suspicious...

https://github.com/tweag/nickel/issues/1930 seems to be the culprit (possibly also the real culprit for https://github.com/tweag/nickel/issues/1630)

**Comment by thufschmitt** [comment]
Rebased on top of #205 . Somehow, this sidesteps the performance issue :partying_face:

**Comment by thufschmitt** [comment]
(Note that this accidentally includes https://github.com/nickel-lang/organist/pull/210. I can revert it if you need @yannham )

File: lib/lib.nix [hunk 3]
```diff
@@ -169,23 +169,19 @@
...
-          ${nickel}/bin/nickel export eval.ncl --field config.flake > $out
-        ''
+      then ''
+        cp -r "${sources}" sources
+        if [ -f sources/nickel.lock.ncl ]; then
+          chmod +w sources sources/nickel.lock.ncl
+        else
+          chmod +w sources
+        fi
+        cp $expectedLockfileContentsPath sources/nickel.lock.ncl
```
Comment (yannham) [inline lib/lib.nix:179]:
```suggestion
        cp "$expectedLockfileContentsPath" sources/nickel.lock.ncl
```

File: lib/files.ncl [hunk 4]
```diff
@@ -43,17 +53,17 @@ let regenerate_files | Files -> nix.derivation.Derivation = fun files_to_generat
...
         nix-s%"
-            rm -f %{target}
-            echo "Regenerating %{target}"
-	    target_dir=$(dirname %{target})
-	    test "${target_dir}" != "." && mkdir -p "${target_dir}"
-            %{copy_command} %{file_in_store} %{target}
+            if [[ ! -f "%{target}" ]] || [[ $(cat "%{target}") != $(cat "%{file_descr.file}") ]]; then
+              rm -f %{target}
+              echo "Regenerating %{target}"
+              target_dir=$(dirname %{target})
```
Comment (yannham) [inline lib/files.ncl:60]:
```suggestion
              target_dir=$(dirname "%{target}")
```

File: lib/files.ncl [hunk 5]
```diff
@@ -43,17 +53,17 @@ let regenerate_files | Files -> nix.derivation.Derivation = fun files_to_generat
...
+              test "${target_dir}" != "." && mkdir -p "${target_dir}"
+              # XXX: If `source.file` is set explicitely to a relative path
+              # and `materialisation_method` is `'Symlink`, this will link to the
+              # original file, not one in the store. Not sure that's what we want.
+              %{copy_command} %{file_descr.file} %{target}
```
Comment (yannham) [inline lib/files.ncl:65]:
```suggestion
              %{copy_command} "%{file_descr.file}" "%{target}"
```

File: lib/files.ncl [hunk 6]
```diff
@@ -83,10 +93,31 @@ let regenerate_files | Files -> nix.derivation.Derivation = fun files_to_generat
...
+      filegen_hook,
+      shells,
       flake.apps.regenerate-files.program = nix-s%"%{regenerate_files files}/bin/regenerate-files"%,
+
+      shells.build.hooks =
+        if filegen_hook.enable then
+          {
+            filegen_hook = nix-s%"
+            %{regenerate_files files}/bin/regenerate-files
+          "%
```
Comment (yannham) [inline lib/files.ncl:118]: The indentation is a bit messed up.
```suggestion
            filegen_hook = nix-s%"
              %{regenerate_files files}/bin/regenerate-files
            "%
```

File: lib/lib.nix (hunk 1 above)
Comment (yannham) [reply lib/lib.nix:214]: Has the doc been lost in a force push?I can't see it still

**Review by yannham — APPROVED** [review]

**Pull request merged**